### PR TITLE
Implement select() and add optional timeout to channel ops

### PR DESCRIPTION
**Pull request by omerfirmak**

### CODE REVIEW

File: source/geod24/concurrency.d [hunk 1]
```diff
@@ -1622,6 +1802,40 @@ final public class Channel (T)
         return false;
     }
 
+    /***********************************************************************
+
+        Try to write a message to Channel without blocking and if it fails,
+        create a write queue entry using the given `sel_state` and `sel_id`
+
+        Param:
+            val = Message to write to channel
```
Comment (AndrejMitrovic) [inline source/geod24/concurrency.d]: val => ptr

File: source/geod24/concurrency.d [hunk 2]
```diff
@@ -1715,6 +1947,40 @@ final public class Channel (T)
         return true;
     }
 
+    /***********************************************************************
+
+        Try to read a message from Channel without blocking and if it fails,
+        create a read queue entry using the given `sel_state` and `sel_id`
+
+        Param:
+            val = Buffer to write the message to
```
Comment (AndrejMitrovic) [inline source/geod24/concurrency.d]: val => ptr

File: source/geod24/concurrency.d [hunk 3]
```diff
@@ -2126,3 +2452,177 @@ public:
         }
     );
 }
+
+@system unittest
```
Comment (AndrejMitrovic) [inline source/geod24/concurrency.d:2491]: `@system` is already the default though.

File: source/geod24/concurrency.d [hunk 4]
```diff
@@ -1497,6 +1532,181 @@ private:
...
+
+        Default ctor
+
+        Params:
+            selectable = A selectable interface reference
+            select_data = pointer to the data for the select operation
+
+    ***********************************************************************/
+
+    this (Selectable selectable, void* select_data)
```
Comment (Geod24) [inline source/geod24/concurrency.d]: `@safe pure nothrow @nogc`

File: source/geod24/concurrency.d [hunk 5]
```diff
@@ -1497,6 +1532,181 @@ private:
...
+
+    Return:
+        Returns success/failure status of the operation and the index
+        of the `Channel` that the operation was carried on. The index is
+        the position of the SelectEntry in `read_list ~ write_list`, ie
+        concatenated lists.
+
+***********************************************************************/
+
+public SelectReturn select (ref SelectEntry[] read_list, ref SelectEntry[] write_list)
```
Comment (Geod24) [inline source/geod24/concurrency.d:1613]: Notes for the future:
Having a `randomShuffle` here is a bit meh. Having mutation on the parameter for the benefit of the function is a red flag. But let's go ahead for now.

Comment (AndrejMitrovic) [reply source/geod24/concurrency.d:1613]: `randomCover` could work?

Comment (omerfirmak) [reply source/geod24/concurrency.d:1613]: `randomCover` wont work. `SelectEntry` s get ID'ed in the order they are traversed.

File: source/geod24/concurrency.d (hunk 3 above)
Comment (omerfirmak) [reply source/geod24/concurrency.d:2491]: Most of the tests in `concurrency.d` already have it. Lets keep it?

Comment (Geod24) [reply source/geod24/concurrency.d:2491]: Yeah

**Comment by omerfirmak** [comment]
All comments addressed

**Review by Geod24 — APPROVED** [review]

**Pull request merged**